File: simulation/routing.py
import heapq
from typing import List, Tuple

def find_shortest_path(manager, start_name: str, end_name: str) -> Tuple[float, List[str]]:
    """
    Dijkstra Algoritması kullanarak iki durak (node) arasındaki en kısa mesafeyi 
    ve geçilmesi gereken durakların sıralı listesini hesaplar.
    """
    cache_key = (start_name, end_name)
    if hasattr(manager, '_path_cache') and cache_key in manager._path_cache:
        return manager._path_cache[cache_key]

    if start_name not in manager.city_map.vertices or end_name not in manager.city_map.vertices:
        return float('inf'), []

    distances = {name: float('inf') for name in manager.city_map.vertices}
    distances[start_name] = 0
    previous_vertices = {name: None for name in manager.city_map.vertices}

    priority_queue = []
    heapq.heappush(priority_queue, (0, start_name))

    while priority_queue:
        current_distance, current_vertex_name = heapq.heappop(priority_queue)

        if current_distance > distances[current_vertex_name]:
            continue

        if current_vertex_name == end_name:
            break

        current_vertex = manager.city_map.vertices[current_vertex_name]
        for edge in current_vertex.adjacencies:
            neighbor_name = edge.target.name
            distance = current_distance + edge.weight
            
            if distance < distances[neighbor_name]:
                distances[neighbor_name] = distance
                previous_vertices[neighbor_name] = current_vertex_name
                heapq.heappush(priority_queue, (distance, neighbor_name))

    path = []
    current = end_name
    
    while current is not None:
        path.append(current)
        current = previous_vertices[current]
    
    path.reverse()

    if distances[end_name] == float('inf'):
        res = float('inf'), []
    else:
        res = distances[end_name], path
        
    if hasattr(manager, '_path_cache'):
        manager._path_cache[cache_key] = res
        
    return res


def find_k_shortest_paths(manager, start_name: str, end_name: str, k: int = 3) -> List[Tuple[float, List[str]]]:
    """
    K-En Kısa Yol (K-Shortest Paths) algoritması (Yen's Algorithm).
    Belirtilen iki düğüm arasında k adet en iyi alternatif rotayı bulur.
    """
    if start_name not in manager.city_map.vertices or end_name not in manager.city_map.vertices:
        return []

    dist, path = manager.find_shortest_path(start_name, end_name)
    if dist == float('inf') or not path:
        return []

    A = [(dist, path)]
    B = []

    for i in range(1, k):
        for j in range(len(A[i-1][1]) - 1):
            spur_node = A[i-1][1][j]
            root_path = A[i-1][1][:j+1]

            removed_edges = []
            
            for _, path_a in A:
                if len(path_a) > j and path_a[:j+1] == root_path:
                    u = path_a[j]
                    v = path_a[j+1]
                    u_vertex = manager.city_map.vertices[u]
                    
                    for edge in list(u_vertex.adjacencies):
                        if edge.target.name == v:
                            u_vertex.adjacencies.remove(edge)
                            removed_edges.append((u, edge))

            removed_nodes = []
            for node in root_path[:-1]:
                if node in manager.city_map.vertices:
                    vertex = manager.city_map.vertices[node]
                    removed_nodes.append((node, vertex.adjacencies))
                    vertex.adjacencies = []

            spur_dist, spur_path = manager.find_shortest_path(spur_node, end_name)

            for node, adjs in removed_nodes:
                manager.city_map.vertices[node].adjacencies = adjs
            for u, edge in removed_edges:
                manager.city_map.vertices[u].adjacencies.append(edge)

            if spur_dist != float('inf') and spur_path:
                total_path = root_path[:-1] + spur_path
                
                total_dist = 0.0
                for m in range(len(total_path) - 1):
                    d, _ = manager.find_shortest_path(total_path[m], total_path[m+1])
                    total_dist += d
                
                if total_path not in [p for _, p in A] and total_path not in [p for _, p in B]:
                    B.append((total_dist, total_path))

        if not B:
            break

        B.sort(key=lambda x: x[0])
        A.append(B.pop(0))

    return A

File: backend/src/main/java/com/ottostop/backend/simulation/Routing.java
package com.ottostop.backend.simulation;

import com.ottostop.backend.models.Edge;
import com.ottostop.backend.models.Vertex;

import java.util.*;

public class Routing {

    public static PathResult findShortestPath(SimulationManager manager, String startName, String endName) {
        String cacheKey = startName + "->" + endName;
        if (manager.getPathCache().containsKey(cacheKey)) {
            return (PathResult) manager.getPathCache().get(cacheKey);
        }

        if (!manager.getCityMap().getVertices().containsKey(startName) || !manager.getCityMap().getVertices().containsKey(endName)) {
            return new PathResult(Double.POSITIVE_INFINITY, new ArrayList<>());
        }

        Map<String, Double> distances = new HashMap<>();
        Map<String, String> previousVertices = new HashMap<>();

        for (String name : manager.getCityMap().getVertices().keySet()) {
            distances.put(name, Double.POSITIVE_INFINITY);
            previousVertices.put(name, null);
        }
        distances.put(startName, 0.0);

        PriorityQueue<NodeDistance> priorityQueue = new PriorityQueue<>(Comparator.comparingDouble(nd -> nd.distance));
        priorityQueue.add(new NodeDistance(startName, 0.0));

        while (!priorityQueue.isEmpty()) {
            NodeDistance current = priorityQueue.poll();
            String currentVertexName = current.nodeName;
            double currentDistance = current.distance;

            if (currentDistance > distances.get(currentVertexName)) {
                continue;
            }

            if (currentVertexName.equals(endName)) {
                break;
            }

            Vertex currentVertex = manager.getCityMap().getVertices().get(currentVertexName);
            for (Edge edge : currentVertex.getAdjacencies()) {
                String neighborName = edge.getTarget().getName();
                double distance = currentDistance + edge.getWeight();

                if (distance < distances.get(neighborName)) {
                    distances.put(neighborName, distance);
                    previousVertices.put(neighborName, currentVertexName);
                    priorityQueue.add(new NodeDistance(neighborName, distance));
                }
            }
        }

        List<String> path = new ArrayList<>();
        String current = endName;

        while (current != null) {
            path.add(current);
            current = previousVertices.get(current);
        }

        Collections.reverse(path);

        PathResult res;
        if (distances.get(endName) == Double.POSITIVE_INFINITY) {
            res = new PathResult(Double.POSITIVE_INFINITY, new ArrayList<>());
        } else {
            res = new PathResult(distances.get(endName), path);
        }

        manager.getPathCache().put(cacheKey, res);
        return res;
    }

    public static List<PathResult> findKShortestPaths(SimulationManager manager, String startName, String endName, int k) {
        if (!manager.getCityMap().getVertices().containsKey(startName) || !manager.getCityMap().getVertices().containsKey(endName)) {
            return new ArrayList<>();
        }

        PathResult shortest = findShortestPath(manager, startName, endName);
        if (shortest.getDistance() == Double.POSITIVE_INFINITY || shortest.getPath().isEmpty()) {
            return new ArrayList<>();
        }

        List<PathResult> A = new ArrayList<>();
        A.add(shortest);
        List<PathResult> B = new ArrayList<>();

        for (int i = 1; i < k; i++) {
            for (int j = 0; j < A.get(i - 1).getPath().size() - 1; j++) {
                String spurNode = A.get(i - 1).getPath().get(j);
                List<String> rootPath = new ArrayList<>(A.get(i - 1).getPath().subList(0, j + 1));

                List<RemovedEdge> removedEdges = new ArrayList<>();

                for (PathResult pathResultA : A) {
                    List<String> pathA = pathResultA.getPath();
                    if (pathA.size() > j && pathA.subList(0, j + 1).equals(rootPath)) {
                        String u = pathA.get(j);
                        String v = pathA.get(j + 1);
                        Vertex uVertex = manager.getCityMap().getVertices().get(u);

                        Iterator<Edge> iterator = uVertex.getAdjacencies().iterator();
                        while (iterator.hasNext()) {
                            Edge edge = iterator.next();
                            if (edge.getTarget().getName().equals(v)) {
                                removedEdges.add(new RemovedEdge(u, edge));
                                iterator.remove();
                            }
                        }
                    }
                }

                List<RemovedNode> removedNodes = new ArrayList<>();
                for (int m = 0; m < rootPath.size() - 1; m++) {
                    String node = rootPath.get(m);
                    if (manager.getCityMap().getVertices().containsKey(node)) {
                        Vertex vertex = manager.getCityMap().getVertices().get(node);
                        removedNodes.add(new RemovedNode(node, new ArrayList<>(vertex.getAdjacencies())));
                        vertex.setAdjacencies(new ArrayList<>());
                    }
                }

                // Temporary clear cache to force recalculation without removed edges
                Map<String, Object> oldCache = new HashMap<>(manager.getPathCache());
                manager.getPathCache().clear();

                PathResult spurResult = findShortestPath(manager, spurNode, endName);

                manager.getPathCache().clear();
                manager.getPathCache().putAll(oldCache);

                // Restore
                for (RemovedNode rn : removedNodes) {
                    manager.getCityMap().getVertices().get(rn.nodeName).setAdjacencies(rn.adjacencies);
                }
                for (RemovedEdge re : removedEdges) {
                    manager.getCityMap().getVertices().get(re.u).getAdjacencies().add(re.edge);
                }

                if (spurResult.getDistance() != Double.POSITIVE_INFINITY && !spurResult.getPath().isEmpty()) {
                    List<String> totalPath = new ArrayList<>(rootPath.subList(0, rootPath.size() - 1));
                    totalPath.addAll(spurResult.getPath());

                    double totalDist = 0.0;
                    for (int m = 0; m < totalPath.size() - 1; m++) {
                        PathResult segRes = findShortestPath(manager, totalPath.get(m), totalPath.get(m + 1));
                        totalDist += segRes.getDistance();
                    }

                    boolean inA = false;
                    for (PathResult pA : A) {
                        if (pA.getPath().equals(totalPath)) { inA = true; break; }
                    }
                    boolean inB = false;
                    for (PathResult pB : B) {
                        if (pB.getPath().equals(totalPath)) { inB = true; break; }
                    }

                    if (!inA && !inB) {
                        B.add(new PathResult(totalDist, totalPath));
                    }
                }
            }

            if (B.isEmpty()) {
                break;
            }

            B.sort(Comparator.comparingDouble(PathResult::getDistance));
            A.add(B.remove(0));
        }

        return A;
    }

    private static class NodeDistance {
        String nodeName;
        double distance;
        NodeDistance(String nodeName, double distance) {
            this.nodeName = nodeName;
            this.distance = distance;
        }
    }

    private static class RemovedEdge {
        String u;
        Edge edge;
        RemovedEdge(String u, Edge edge) {
            this.u = u;
            this.edge = edge;
        }
    }

    private static class RemovedNode {
        String nodeName;
        List<Edge> adjacencies;
        RemovedNode(String nodeName, List<Edge> adjacencies) {
            this.nodeName = nodeName;
            this.adjacencies = adjacencies;
        }
    }
}
